go/ai/generate.go

```diff
@@ -86,8 +86,122 @@ func LookupModel(provider, name string) *Model {
 	return (*Model)(core.LookupActionFor[*GenerateRequest, *GenerateResponse, *GenerateResponseChunk](atype.Model, provider, name))
 }
 
-// Generate applies the [Model] to some input, handling tool requests.
-func (m *Model) Generate(ctx context.Context, req *GenerateRequest, cb ModelStreamingCallback) (*GenerateResponse, error) {
+// GenerateRequestBuilder adds
+type GenerateRequestBuilder func(req *GenerateRequest)
+
+// WithTextPrompt adds a simple text user prompt to GenerateRequest
+func WithTextPrompt(prompt string) GenerateRequestBuilder {
+	return func(req *GenerateRequest) {
+		req.Messages = append(req.Messages, NewUserTextMessage(prompt))
+	}
+}
+
+// WithSystemPrompt adds a simple text system prompt to GenerateRequest
+func WithSystemPrompt(prompt string) GenerateRequestBuilder {
+	return func(req *GenerateRequest) {
+		req.Messages = append(req.Messages, NewSystemTextMessage(prompt))
+	}
+}
+
+// WithMessages adds provided messages to GenerateRequest
+func WithMessages(messages ...*Message) GenerateRequestBuilder {
+	return func(req *GenerateRequest) {
+		req.Messages = append(req.Messages, messages...)
+	}
+}
+
+// WithConfig adds provided config to GenerateRequest
+func WithConfig(config any) GenerateRequestBuilder {
+	return func(req *GenerateRequest) {
+		req.Config = config
+	}
+}
+
+// WithConfig adds provided candidate count to GenerateRequest
+func WithCandidates(c int) GenerateRequestBuilder {
+	return func(req *GenerateRequest) {
+		req.Candidates = c
+	}
+}
+
+// WithConfig adds provided context to GenerateRequest
+func WithContext(c ...any) GenerateRequestBuilder {
+	return func(req *GenerateRequest) {
+		req.Context = c
+	}
+}
+
+// WithConfig adds provided tools to GenerateRequest
+func WithTools(tools ...Tool) GenerateRequestBuilder {
+	return func(req *GenerateRequest) {
+		var toolDefs []*ToolDefinition
+		for _, t := range tools {
+			toolDefs = append(toolDefs, t.Definition())
+		}
+		req.Tools = toolDefs
+	}
+}
+
+// WithConfig adds provided output schema to GenerateRequest
+func WithOutputSchema[T any](schema T) GenerateRequestBuilder {
+	return func(req *GenerateRequest) {
+		req.Output.Schema = base.SchemaAsMap(base.InferJSONSchema(schema))
+	}
+}
+
+// WithConfig adds provided output format to GenerateRequest
+func WithOutputFormat(format OutputFormat) GenerateRequestBuilder {
+	return func(req *GenerateRequest) {
+		req.Output.Format = format
+	}
+}
+
+// Generate run generate request for this model. Returns GenerateResponse struct.
+func (m *Model) Generate(ctx context.Context, withs ...GenerateRequestBuilder) (*GenerateResponse, error) {
+	return m.StreamGenerate(ctx, nil, withs...)
+}
+
+// GenerateText run generate request for this model. Returns generated text only.
+func (m *Model) GenerateText(ctx context.Context, withs ...GenerateRequestBuilder) (string, error) {
```

Comment:
Should this return `(string, *GenerateResponse, error)`? You can always ignore the full response if you want with a `_`, but that way you still have access to the response if you want to get metadata out of it.

Reply:
Muddies the water with plain `Generate` a bit...

Reply:
Yeah, I think I'd be okay with `resp.Text()` being a non-erroring method (it should be - nothing can go wrong at that point) and not returning the response from GenerateText

Reply:
Done -- `resp.Text()` doesn't error anymore